utils/cp/CPUtils.cpp

```diff
@@ -0,0 +1,170 @@
+        switch (previous_state) {
+        case types::cb_board_support::CPState::A:
+            return types::cb_board_support::CPState::A;
+        default:
+            return types::cb_board_support::CPState::B;
+        }
+    }
+
+    if (voltage >= 8000 /* mV */) /* 8 V <= x < 10 V */
+        return types::cb_board_support::CPState::B;
+
+    if (voltage >= 7000 /* mV */) { /* 7 V <= x < 8 V */
+        switch (previous_state) {
+        case types::cb_board_support::CPState::A:
+        case types::cb_board_support::CPState::B:
+            return types::cb_board_support::CPState::B;
+        default:
+            return types::cb_board_support::CPState::C;
+        }
+    }
+
+    if (voltage >= 5000 /* mV */) /* 5 V <= x < 7 V */
+        return types::cb_board_support::CPState::C;
+
+    if (voltage >= 4000 /* mV */) { /* 4 V <= x < 5 V */
+        switch (previous_state) {
+        case types::cb_board_support::CPState::A:
+        case types::cb_board_support::CPState::B:
+        case types::cb_board_support::CPState::C:
+            return types::cb_board_support::CPState::C;
+        default:
+            return types::cb_board_support::CPState::D;
+        }
+    }
+
+    if (voltage >= CP_STATE_D_LOWER_THRESHOLD) /* 2 V <= x < 4 V */
+        return types::cb_board_support::CPState::D;
+
+    if (voltage >= -1000 /* mV */) /* -1 V <= x < 2 V */
+        return types::cb_board_support::CPState::E;
+
+    if (voltage >= -2000 /* mV */) { /* -2 V <= x < -1 V (not standard) */
+        switch (previous_state) {
+        case types::cb_board_support::CPState::A:
+        case types::cb_board_support::CPState::B:
+        case types::cb_board_support::CPState::C:
+        case types::cb_board_support::CPState::D:
+        case types::cb_board_support::CPState::E:
+            return types::cb_board_support::CPState::E;
+        default:
+            return types::cb_board_support::CPState::PilotFault;
+        }
+    }
+
+    if (voltage >= -10000 /* mV */) /* -10 V <= x < -2 V */
+        return types::cb_board_support::CPState::PilotFault;
+
+    if (voltage >= -11000 /* mV */) { /* -11 V <= x < -10 V (not standard) */
+        return previous_state == types::cb_board_support::CPState::F ? types::cb_board_support::CPState::F
+                                                                     : types::cb_board_support::CPState::PilotFault;
+    }
+
+    if (voltage >= -13000 /* mV */) /* -13 V <= x < -11 V */
+        return types::cb_board_support::CPState::F;
+
+    return types::cb_board_support::CPState::PilotFault; /* < -13 V */
+}
+
+bool CPUtils::check_for_cp_errors(cp_state_errors& cp_errors,
+                                  types::cb_board_support::CPState& current_cp_state,
+                                  const double& duty_cycle,
+                                  const int& voltage_neg_side,
+                                  const int& voltage_pos_side) {
```

Comment:
I'm not really happy about this design of this function. It's very opaque and complex. Is it possible to split?
is_cp_short_fault()
is_ventilation_fault()
is_diode_fault()
is_pilot_fault()

This should make unittest more easier.